### PR TITLE
Emit specific message for time<=0.3.35

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0282]: type annotations needed for `Box<_>`
  --> /home/gh-estebank/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.34/src/format_description/parse/mod.rs:83:9
   |
83 |     let items = format_items
   |         ^^^^^
...
86 |     Ok(items.into())
   |              ---- type must be known at this point
   |
   = note: this is an inference error on `time` caused by a change in Rust 1.80.0; update `time` to version `>=0.3.36`
```

Partially mitigate the fallout from https://github.com/rust-lang/rust/issues/127343. Although the biggest benefit of this would have been if we had had this in 1.80 before it became stable, the long-tail of that change will be felt for a *long* time, so better late than never.

We can also emit an even more targeted error instead of this inference failure.

### CODE REVIEW

**Comment by estebank** [comment]
This PR is missing a test because it relies on cargo paths and synthesizing that in a make will still be ugly. I verified with a minimal local project:

`$ cargo new time-hack`
`Cargo.toml`:
```
[package]
name = "time-hack"
version = "0.1.0"
edition = "2021"

[dependencies]
time = { version = "=0.3.34", features = ["parsing"] }
```

The output is the one in the PR description.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:664]: Hard-coding an error message like this based off of path matching can likely be written using regex instead. This let chain is inscrutable to the point that I'd rather we not merge this if it can't be rewritten.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:624]: Please leave a FIXME to remove this in a few versions when the time-0.3.35 and below's downloads are lower.

I very much do not want this hack to remain in the compiler for very long. IMO all crate-related hacks in the compiler (even the very old ones, like the rental and bevy hacks) should have some sort of "phasing out" condition.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:664]: Or, alternatively, I'd rather we not do path parsing here and just match on a combination of crate name and some other condition we can deduce from the program itself.

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:664]: like, are we even always going to be compiling `time` from a crates.io registry? Does this work correctly for vendored deps? This certainly doesn't work for non-cargo build tools, right?

Comment (estebank) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:664]: Pattern matching on a path sounds more brittle. This check is horrible, I agree, but it at least handles paths across windows/unix correctly, and by being this *specific* in the number of checks we're less likely to wrongly trigger if someone just happens to put code in a directory called `time-0.blah`. I think we can likely split some of the elements into one or two closures to make the chain easier to read, though.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (estebank) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:624]: We have discussed having these kind of checks being done by cargo using some kind of database/file mechanism a-la RustSect for a while. I think that's got to be the future of these (which is why I mention it in the FIXME). Once we do that we no longer have to keep these kind of hacks in the compiler itself and it gets divorced from how many downloads the crate has (trying to compile a project that uses `time = "0.3.34"` 5 years from now, I'd appreciate having *some* hint of what's broken).

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:664]: This *is* already incredibly brittle. See that comment I just left. I'd rather we not make this hack rely on some special structure of the crates.io registry, or not go with this at all if that's impossible.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:624]: This is not the first nor last time that we have broken a dependency due to an impl, and honestly I'd rather we *not* special case `time`. Nobody is committing to the follow-up work of cleaning up this hack, and IMO this simply leaves the compiler with one more special case that will continue to bitrot as time goes on.

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:624]: The fact that we, as T-compiler, need to maintain an ever-growing set of diagnostics hacks for things like this makes me pretty uncomfortable, expecially because this:

> We have discussed having these kind of checks being done by cargo using some kind of database/file mechanism a-la RustSect for a while.

...is just an *idea*, so we cannot be relying it on some future possibility as the removal condition of this hack.

**Comment by compiler-errors** [comment]
I also don't see why this isn't something that **cargo** should be detecting anytime that a registry dependency fails to compile.

Anytime we have a registry dep fail to compile, it should probably be saying something like "hey, try updating the version of crate xyz you depend on", since the fact that it was both published to the registry but now fails to compile suggests that there's probably something that happened in the mean time.

These "somethings" could include:
* Inference failures like this
* Soundness fixes
* Future-compat warnings that were turned into hard errors
* Accidental stabilizations that were rolled back

There are many reasons why a crate compiles on rust 1.x but not 1.x+1 that we regularly accept for one reason or the other, and I still don't think that we should or need to have a crate-specific hack for time in this case.

**Comment by compiler-errors** [comment]
If you disagree, though, then I guess you probably should compiler-nominate this for team discussion, because I would opt to not approve this now, but I'd rather not be the single blocker for this hack.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:657]: This can technically now have a UI test written for it, so it should probably.

**Comment by estebank** [comment]
> I also don't see why this isn't something that cargo should be detecting any time that a registry dependency fails to compile.

I agree, our user experience around updates doesn't live up to the bar we set for ourselves.

> There are many reasons why a crate compiles on rust 1.x but not 1.x+1 that we regularly accept for one reason or the other, and I still don't think that we should or need to have a crate-specific hack for time in this case.

If it was up to me, we'd have 1.80.2 without the new `impl` yesterday. This was a fuck up and we *need* to do something to mitigate the fallout as much as we can. 5% of all of crates.io broke, and we didn't even bother mitigating the fallout. An explicit check for this error is the *very least* we should do, and we can't wait for a full design-discussion-implementation-stabilization cycle of a cargo feature to that "right".

**Comment by compiler-errors** [comment]
i'll let someone else review this after the team decides on the action here

r? compiler

**Comment by jieyouxu** [comment]
I appreciate the efforts to help mitigate the fallout from the inference regression, but I'm also not very comfortable adding another crate-specific hack. If anything, we ought to be *removing* crate-specific hacks. I would prefer if T-compiler worked with T-libs-api and T-release to create a postmortem for the time regression to make our release process more robust. To help catch these inference regressions happen *before* they happen in the future, instead of adding hacks after the fact. Since this has been nominated for T-compiler decision, I'll wait for that decision and respect that decision, even though I would prefer not having this hack myself.

**Comment by jieyouxu** [comment]
In any case, I'll write up a summary for T-compiler triage meeting.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (jieyouxu) [inline compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:632]: Question: with this new detection logic, AFAICT we're not actually checking for specific `time` versions (which is impossible to check in the general case without assuming things about cargo and such). If that's the case, is it possible that the note

```
this is an inference error on crate `time` caused by a change in Rust 1.80.0; update `time` to version `>=0.3.36`
```

sounds too certain? i.e. can this also be caused by a not-caused-by-us inference error due to `time` itself (in a newer version)?

**Comment by jieyouxu** [comment]
### Summary for T-compiler

This PR proposes to special-case a diagnostic note for the `time` crate to help mitigate ecosystem fallout from https://github.com/rust-lang/rust/issues/127343:

```
error[E0282]: type annotations needed for `Box<Vec<_>>`
  --> $DIR/detect-old-time-version-format_description-parse.rs:4:9
   |
LL |     let items = Box::new(vec![]);
   |         ^^^^^   ---------------- type must be known at this point
   |
   = note: this is an inference error on crate `time` caused by a change in Rust 1.80.0; update `time` to version `>=0.3.36`
```

Nominating for T-compiler discussion to decide if we want to merge this crate-specific diagnostic note.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (yotamofek) [inline compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:657]: This could be made much nicer (IMHO) with let chains:

```suggestion
        if self.infcx.tcx.crate_name(LOCAL_CRATE) == sym::time
            && let Some(span) = span
            && let InferSourceKind::LetBinding { pattern_name: Some(name), .. } = kind
            && name == "items"
            && let FileName::Real(file) = self.infcx.tcx.sess.source_map().span_to_filename(span)
            && {
                let path = file.local_path_if_available().to_string_lossy();
                path.contains("format_description") && path.contains("parse")
            }
        {
            infer_subdiags.clear();
            Some(())
        } else {
            None
        }
```

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (estebank) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:632]: If the coincidence of all of these conditions happening were to happen, it would mainly confuse the `time` maintainers and no-one else :)

We can also look at the path again and change the confidence of the wording depending on whether we find the version from the directory name or not (or try to find the `Cargo.toml`), but I'd rather we didn't do that (or if we start doing that, have a single place where we maintain a "get crate version from path" logic, as there *are* a bunch of places where it'd be useful to know the crate version for a given type, but I feel all of this is beyond the scope of "get something in front of users quickly").

Comment (jieyouxu) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:632]: Yeah I'd rather not probe for `Cargo.toml` since that's not really our concern (as in rustc) I feel like 💀. I'm okay with accepting this version of detection logic if the team decides to.

**Comment by m-ou-se** [comment]
> 5% of all of crates.io broke

I don't agree with this framing. Cargo does not use the lock files from packages on crates.io by default. So if you pull anything from crates.io with cargo, it'll simply use the newest (semver compatible) version of the `time` crate, which compiles fine.

People working in projects with a Cargo.lock with an old `time` crate file are broken, yes (until they run cargo update), but that's not the same as 'packages on crates.io broke'.

> If it was up to me, we'd have 1.80.2 without the new impl yesterday

That would be a breaking change too. That would break anyone who uses the new impls in 1.80.0 when they update their compiler again, but without a "cargo update" solution.

**Comment by apiraino** [comment]
Discussed in t-compiler triage [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-08-22/near/464395365).

This change is a bit controversial, we don't like hard-coding stuff for specific crates but it seems it's [not](https://github.com/rust-lang/rust/issues/73933) the [first](https://github.com/rust-lang/rust/issues/83125) [time](https://github.com/rust-lang/rust/issues/119956)[^1] and after all it's just a diagnostic message. That we pinky promise to remove in around six months or so. I hereby @apiraino take responsability to remind t-compiler in case the revert is forgotten :slightly_smiling_face: 

@rustbot label -I-compiler-nominated

---

[^1]: see what I did there?

**Comment by estebank** [comment]
Added a `const fn` with a `version_check` so that we *can't* forget :)

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (jhpratt) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:632]: > If the coincidence of all of these conditions happening were to happen, it would mainly confuse the time maintainers and no-one else :)

For certainty, I'm fine with that :slightly_smiling_face:

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (jieyouxu) [inline compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:641]: Suggestion: thanks, I do like this technique so we *can't* possibly forgor about this. Could you please also add a `FIXME(#129461)` so it's not only present in the code?

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (jieyouxu) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:666]: Discussion (for @estebank): I think this does make it easier to follow what conditions are needed to trigger the diagnostic, but I'm not going to block on that. At your discretion, whatever is easier to read for you.

File: tests/ui/inference/detect-old-time-version-format_description-parse.rs
Comment (jieyouxu) [inline tests/ui/inference/detect-old-time-version-format_description-parse.rs:1]: Remark: *technically* I believe we would also trigger the diagnostic for any crates named `time` (not necessarily *the* `time` crate on crates.io), but as rustc for our intents and purposes we can't know in general, so I think it's fine.

File: compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
Comment (jieyouxu) [reply compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs:625]: Resolving as this is now implemented under a hard time-gate which will block compilation if it's not removed before bumping to 1.89.

**Review by jieyouxu — APPROVED** [review]
Thanks, this looks good to me in general. I have a suggestion to link to a specific issue (opened a new one at #129461) so it's not just present in code, and a couple of minor nits.

Please feel free to r=me after that.

**Comment by jieyouxu** [comment]
Also nominating this for beta backport as noted in the compiler triage meeting (https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-08-22/near/464401318).

@rustbot +beta-nominated

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-08-29/near/466023240). A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +beta-accepted

**Pull request merged**
